File: acorn-core/src/test/java/com/ardley/acorn/action/ActionRegistryTest.java
package com.ardley.acorn.action;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

/**
 * Tests action registration, resolution, and duplicate detection.
 */
class ActionRegistryTest {

    enum TestActions implements Action {
        ListUsers("List all users"),
        UpdateUser("Update a user"),
        DeleteUser("Delete a user");

        private final String desc;
        TestActions(String desc) { this.desc = desc; }

        // Enum.name() already satisfies Action.name() — no override needed
        @Override public String description() { return desc; }
    }

    @Test
    @DisplayName("Registers and resolves actions by name")
    void registerAndResolve() {
        ActionRegistry registry = new ActionRegistry();
        registry.register(TestActions.ListUsers);
        registry.register(TestActions.UpdateUser);

        Action resolved = registry.resolve("ListUsers");

        assertThat(resolved.name()).isEqualTo("ListUsers");
        assertThat(resolved.description()).isEqualTo("List all users");
    }

    @Test
    @DisplayName("registerAll registers entire enum")
    void registerAllEnum() {
        ActionRegistry registry = new ActionRegistry();
        registry.registerAll(TestActions.class);

        assertThat(registry.size()).isEqualTo(3);
        assertThat(registry.resolve("DeleteUser").description()).isEqualTo("Delete a user");
    }

    @Test
    @DisplayName("Resolving unknown action throws with descriptive message")
    void resolveUnknownThrows() {
        ActionRegistry registry = new ActionRegistry();

        assertThatThrownBy(() -> registry.resolve("NonExistent"))
                .isInstanceOf(IllegalArgumentException.class)
                .hasMessageContaining("Unknown action")
                .hasMessageContaining("NonExistent");
    }

    @Test
    @DisplayName("Duplicate registration throws immediately")
    void duplicateRegistrationThrows() {
        ActionRegistry registry = new ActionRegistry();
        registry.register(TestActions.ListUsers);

        assertThatThrownBy(() -> registry.register(TestActions.ListUsers))
                .isInstanceOf(IllegalArgumentException.class)
                .hasMessageContaining("Duplicate action name")
                .hasMessageContaining("ListUsers");
    }

    @Test
    @DisplayName("all() returns unmodifiable collection of registered actions")
    void allReturnsRegistered() {
        ActionRegistry registry = new ActionRegistry();
        registry.registerAll(TestActions.class);

        assertThat(registry.all())
                .hasSize(3)
                .extracting(Action::name)
                .containsExactlyInAnyOrder("ListUsers", "UpdateUser", "DeleteUser");
    }
}

File: node/packages/core/src/__tests__/action-registry.test.ts
import { describe, test, expect } from 'bun:test';
import { ActionRegistry } from '../action-registry.js';
import { defineActions } from '../define-actions.js';

describe('ActionRegistry', () => {
  test('register and resolve', () => {
    const registry = new ActionRegistry();
    registry.register({ name: 'ListUsers', description: 'List users' });

    const action = registry.resolve('ListUsers');
    expect(action.name).toBe('ListUsers');
    expect(action.description).toBe('List users');
  });

  test('throws on duplicate', () => {
    const registry = new ActionRegistry();
    registry.register({ name: 'ListUsers', description: 'List users' });
    expect(() => registry.register({ name: 'ListUsers', description: 'dup' })).toThrow('Duplicate');
  });

  test('throws on unknown action', () => {
    const registry = new ActionRegistry();
    expect(() => registry.resolve('Unknown')).toThrow('Unknown action');
  });

  test('registerAll from defineActions', () => {
    const actions = defineActions({
      ListUsers: 'List users',
      UpdateUser: 'Update user',
      DeleteUser: 'Delete user',
    });

    const registry = new ActionRegistry();
    registry.registerAll(actions);

    expect(registry.size()).toBe(3);
    expect(registry.resolve('UpdateUser').description).toBe('Update user');
  });

  test('all() returns registered actions', () => {
    const registry = new ActionRegistry();
    registry.register({ name: 'A', description: 'a' });
    registry.register({ name: 'B', description: 'b' });

    expect(registry.all().length).toBe(2);
  });
});

describe('defineActions', () => {
  test('derives name from key', () => {
    const actions = defineActions({
      ListUsers: 'List all users',
      UpdateUser: 'Modify a user',
    });

    expect(actions.ListUsers.name).toBe('ListUsers');
    expect(actions.ListUsers.description).toBe('List all users');
    expect(actions.UpdateUser.name).toBe('UpdateUser');
  });

  test('result is frozen', () => {
    const actions = defineActions({ Foo: 'bar' });
    expect(Object.isFrozen(actions)).toBe(true);
    expect(Object.isFrozen(actions.Foo)).toBe(true);
  });
});
